### ISSUE TITLE
Runtime error computing await expression

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/sdk#29228]
The following code should print `FOO`:

```dart
import 'dart:async';

Future<String> foo() async {
  return "foo";
}

Future<String> bar() async {
  return ((await foo()) ?? "bar").toUpperCase();
}

main() async {
  var val = await bar();
  print(val);
}
```

Instead, it's triggering this error:

```
/Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:22343
  throw error;
  ^

TypeError: (intermediate value)(...)[dartx.toUpperCase] is not a function
    at /Users/vsm/ddctest/awaitexpr.js:22:28
    at next (native)
    at onValue (/Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:2041:22)
    at /Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:2059:18
    at _AsyncCallbackEntry.async.scheduleMicrotask.dart.fn.e [as callback] (/Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:18192:29)
    at async._microtaskLoop (/Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:19253:11)
    at async._startMicrotaskLoop (/Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:19260:11)
    at Immediate.internalCallback [as _onImmediate] (/Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:19364:7)
    at tryOnImmediate (timers.js:543:15)
    at processImmediate [as _immediateCallback] (timers.js:523:5)
```

**Comment by vsmenon** [comment]
Note, if I rewrite to:

```
Future<String> bar() async {
  var tmp = ((await foo()) ?? "bar");
  return tmp.toUpperCase();
}
```
it works.

**Comment by vsmenon** [comment]
@jmesserly - this is coming up in at least one internal test.  I'll see if I can workaround it for now.

**Comment by jmesserly** [comment]
err, any idea what's going wrong with it? I can't think why that would matter at all. Maybe our code generation is busted for `??` in some cases? We go through a lot of effort to simplify the resulting let-expression, that code was tested before "await" so could be some issues there.

**Comment by vsmenon** [comment]
Not quite sure.  Here's the generated code for the version that's breaking:
```
awaitexpr.bar = function() {
  return dart.async(function*() {
    return yield* (function*() {
      let l = (yield awaitexpr.foo());
      return l != null ? l : "bar";
    })()[dartx.toUpperCase]();
  }, core.String);
};
```

And with the workaround (adding the temp) - this runs fine:
```
awaitexpr.bar = function() {
  return dart.async(function*() {
    let tmp = yield* (function*() {
      let l = (yield awaitexpr.foo());
      return l != null ? l : "bar";
    })();
    return tmp[dartx.toUpperCase]();
  }, core.String);
};
```

**Comment by vsmenon** [comment]
Is is possibly a precedence issue with `yield*`?

**Comment by jmesserly** [comment]
yup sounds right. This part is wrong:
```dart
    return yield* (function*() {
      let l = (yield awaitexpr.foo());
      return l != null ? l : "bar";
    })()[dartx.toUpperCase]();
```
it should be:
```dart
    return (yield* (function*() {
      let l = (yield awaitexpr.foo());
      return l != null ? l : "bar";
    })())[dartx.toUpperCase]();
```

**Comment by jmesserly** [comment]
fix for that is typically in js_ast nodes.dart and/or printer.dart. FYI.

**Comment by jmesserly** [comment]
Aside, it also seems a little bit busted we generated:
```js
let l = (yield awaitexpr.foo());
```
should be:
```js
let l = yield awaitexpr.foo();
```
might be a symptom of the same problem. However that one is purely cosmetic.

**Comment by vsmenon** [comment]
Yup, verified by adding the paren (by hand) - thanks!  :-)

**Comment by vsmenon** [comment]
Workaround was already landed downstream, so I'll keep this as P2, 1.24.

**Comment by jmesserly** [comment]
https://codereview.chromium.org/2804833002/
